File: Solver.py
"""
by : Luvidi (2021)
f=front b=back u=up d=down r=right l=left
w=white y=yellow b=blue g=green o=orange r=red

---------
| 0 | 1 |
| 3 | 2 |
---------
"""

def replace(string,index,new) :
    return string[:index]+new+string[index+1:]

class Rubik :
    def __init__(self,color) :
        self.color=color
        
    def F(self) :
        temp=self.color[3]
        self.color=replace(self.color,3,self.color[2])
        self.color=replace(self.color,2,self.color[1])
        self.color=replace(self.color,1,self.color[0])
        self.color=replace(self.color,0,temp)

        temp=self.color[11]
        self.color=replace(self.color,11,self.color[22])
        self.color=replace(self.color,22,self.color[13])
        self.color=replace(self.color,13,self.color[16])
        self.color=replace(self.color,16,temp)

        temp=self.color[21]
        self.color=replace(self.color,21,self.color[12])
        self.color=replace(self.color,12,self.color[19])
        self.color=replace(self.color,19,self.color[10])
        self.color=replace(self.color,10,temp)

    def R(self) :
        temp=self.color[19]
        self.color=replace(self.color,19,self.color[18])
        self.color=replace(self.color,18,self.color[17])
        self.color=replace(self.color,17,self.color[16])
        self.color=replace(self.color,16,temp)

        temp=self.color[2]
        self.color=replace(self.color,2,self.color[14])
        self.color=replace(self.color,14,self.color[4])
        self.color=replace(self.color,4,self.color[10])
        self.color=replace(self.color,10,temp)

        temp=self.color[1]
        self.color=replace(self.color,1,self.color[13])
        self.color=replace(self.color,13,self.color[7])
        self.color=replace(self.color,7,self.color[9])
        self.color=replace(self.color,9,temp)

    def U(self) :
        temp=self.color[11]
        self.color=replace(self.color,11,self.color[10])
        self.color=replace(self.color,10,self.color[9])
        self.color=replace(self.color,9,self.color[8])
        self.color=replace(self.color,8,temp)

        temp=self.color[1]
        self.color=replace(self.color,1,self.color[17])
        self.color=replace(self.color,17,self.color[5])
        self.color=replace(self.color,5,self.color[21])
        self.color=replace(self.color,21,temp)

        temp=self.color[0]
        self.color=replace(self.color,0,self.color[16])
        self.color=replace(self.color,16,self.color[4])
        self.color=replace(self.color,4,self.color[20])
        self.color=replace(self.color,20,temp)
    
    def solved(self) :
        for i in range(6) :
            first=self.color[i*4]
            for j in range(4) :
                if self.color[i*4+j]!=first :
                    return False
        return True
    
    def solve(self) : 
        s=self.getMoves()
        s2=""
        col=s[0]
        cnt=0
        for i in range(len(s)) :
            if s[i]==col :
                cnt+=1
            else :
                if cnt==1 :
                    s2+=col+" "
                elif cnt==2 :
                    s2+=col+"2 "
                elif cnt==3 :
                    s2+=col+"' "
                col=s[i]
                cnt=1
        if cnt==1 :
            s2+=col+" "
        elif cnt==2 :
            s2+=col+"2 "
        elif cnt==3 :
            s2+=col+"' "
        return s2
        
        
    def getMoves(self) :
        if self.solved() :
            return ""
        
        original=self.color
        moves1, moves2, list1, list2 = [], [], [], []
        
        for moveIdx in range(3) :
            self.color=original
            move=["F","U","R"][moveIdx]
            exec("self."+move+"()")
            if self.solved() :
                print(self.color)
                return move
            list1.append(self.color)
            moves1.append(move)
        
        while True :
            moves2, list2 = [], []
            for cnt in range(len(moves1)) :
                for moveIdx in range(3) :
                    self.color=list1[cnt]
                    move=["F","U","R"][moveIdx]
                    exec("self."+move+"()")
                    if self.solved() :
                        return moves1[cnt]+move
                    list2.append(self.color)
                    moves2.append(moves1[cnt]+move)
            list1, moves1 = list2.copy(), moves2.copy()
            
        
        
if __name__=='__main__' :
    # f b u d r l
    color="obbygyrgrwowowrwyoyrbbgg" # R F2 R2 F U2 R F2 
    r=Rubik(color) 
    print(r.solve())

File: Solver.cpp
#include <iostream>
#include <string>

using namespace std;

class Rubik {
    public :
        // f b u d r l
        string color; 
            
        char temp;
        
        Rubik(string c) {
            color=c;
        }
    
        void F() {
            temp=color[3];
            color[3]=color[2];
            color[2]=color[1];
            color[1]=color[0];
            color[0]=temp;

            temp=color[11];
            color[11]=color[22];
            color[22]=color[13];
            color[13]=color[16];
            color[16]=temp;

            temp=color[21];
            color[21]=color[12];
            color[12]=color[19];
            color[19]=color[10];
            color[10]=temp;           
        }
        
        void R() {
            temp=color[19];
            color[19]=color[18];
            color[18]=color[17];
            color[17]=color[16];
            color[16]=temp;

            temp=color[2];
            color[2]=color[14];
            color[14]=color[4];
            color[4]=color[10];
            color[10]=temp;

            temp=color[1];
            color[1]=color[13];
            color[13]=color[7];
            color[7]=color[9];
            color[9]=temp;
        }

        void U() {
            temp=color[11];
            color[11]=color[10];
            color[10]=color[9];
            color[9]=color[8];
            color[8]=temp;

            temp=color[1];
            color[1]=color[17];
            color[17]=color[5];
            color[5]=color[21];
            color[21]=temp;

            temp=color[0];
            color[0]=color[16];
            color[16]=color[4];
            color[4]=color[20];
            color[20]=temp;
        }


        bool solved() {
            char first;
            for (int i=0; i<6; i++) {
                first=color[4*i];
                for (int j=1; j<4; j++) {
                    if (color[4*i+j]!=first) {
                        return false;
                    }
                }
            }
            return true;
        }

        string solve() {
            string s=getMoves();
            
            string s2="";
            char col=color[0];
            int cnt=0;
            if (s=="") {
                return "";
            }
            for (int i=0; i<s.length(); i++) {
                if (s[i]==col) {
                    cnt++;
                } 
                else {
                    switch (cnt) {
                        case 1:
                            s2+=col;
                            s2+=' ';
                            break;
                        case 2:
                            s2+=col;
                            s2+="2 ";
                            break;
                        case 3:
                            s2+=col;
                            s2+="' ";
                            break;
                    }
                    col=s[i];
                    cnt=1;
                }
            }
            switch (cnt) {
                case 1:
                    s2+=col;
                    s2+=' ';
                    break;
                case 2:
                    s2+=col;
                    s2+="2 ";
                    break;
                case 3:
                    s2+=col;
                    s2+="' ";
                    break;
                default:
                    s2+=" ";
                    break;
            }
            return s2;
        }

        string getMoves() {
            if (solved()) {
                return "";
            }
            
            string original = color;
            string moves1 = "";
            string list1 = "";
            string moves2 = "";
            string list2 = "";
            string prevMove = "";
            int moveLen = 0;

            for (int move=0; move<3; move++) {
                color=original;
                switch (move) {
                    case 0 :
                        F();
                        if (solved()) {return "F";}
                        list1 += color;
                        moves1 += "F";
                        break;
                    case 1 :
                        U();
                        if (solved()) {return "U";}
                        list1 += color;
                        moves1 += "U";
                        break;
                    case 2 :
                        R();
                        if (solved()) {return "R";}
                        list1 += color;
                        moves1 += "R";
                        break;
                }
            }
            moveLen++;
            
            while (true) {
                moves2 = "";
                list2="";
                for (int cnt=0; cnt<list1.size()/24; cnt++) {
                    for (int move=0; move<3; move++) {
                        color=list1.substr(24*cnt,24);
                        prevMove=moves1.substr(moveLen*cnt,moveLen);
                        switch (move) {
                            case 0 :
                                F();
                                if (solved()) {return prevMove+"F";}
                                list2 += color;
                                moves2 += prevMove+"F";
                                break;
                            case 1 :
                                U();
                                if (solved()) {return prevMove+"U";}
                                list2 += color;
                                moves2 += prevMove+"U";
                                break;
                            case 2 :
                                R();
                                if (solved()) {return prevMove+"R";}
                                list2 += color;
                                moves2 += prevMove+"R";
                                break;
                        }
                    }
                }
                moveLen++;

                moves1 = "";
                list1 = "";
                for (int cnt=0; cnt<list2.size()/24; cnt++) {
                    for (int move=0; move<3; move++) {
                        color=list2.substr(24*cnt,24);
                        prevMove=moves2.substr(moveLen*cnt,moveLen);
                        switch (move) {
                            case 0 :
                                F();
                                if (solved()) {return prevMove+"F";}
                                list1 += color;
                                moves1 += prevMove+"F";
                                break;
                            case 1 :
                                U();
                                if (solved()) {return prevMove+"U";}
                                list1 += color;
                                moves1 += prevMove+"U";
                                break;
                            case 2 :
                                R();
                                if (solved()) {return prevMove+"R";}
                                list1 += color;
                                moves1 += prevMove+"R";
                                break;
                        }
                    }
                }
                moveLen++;
            }        
        }

};

int main() {
    string color = "obbygyrgrwowowrwyoyrbbgg"; // R F2 R2 F U2 R F2 
    Rubik r(color);

    string s=r.solve();

    cout << s << endl;
    return 0;
}